README.md
# nacelle-loox-nuxt-module

Adds Vue.js components for integrating [Loox](https://loox.app/) product reviews in your [Nacelle](https://getnacelle.com/) Nuxt project.

## Requirements

- A Nacelle project set up locally. To get started, refer to [the Nacelle docs](https://docs.getnacelle.com).
- Loox product reviews exist for your Shopify site.

## Setup

### Add Module to Nacelle

Once you have Nacelle set up, you can install this module in your project from `npm`:

```sh
npm install @nacelle/nacelle-loox-nuxt-module --save
```

After the package has installed, open `nuxt.config.js`. Under `modules` add `@nacelle/nacelle-loox-nuxt-module` to the array:

```js
modules: [
  // ...other modules,
  '@nacelle/nacelle-loox-nuxt-module'
],
```

Then add your store's Shopify domain to your environment variables `.env` file:

```sh
# example: "starship-furniture" for "starship-furniture.myshopify.com"
SHOPIFY_DOMAIN=xxxxxxxxxxxxx
```

Lastly, add the environment variable to the `nacelle` block of `nuxt.config.js`:

```javascript
// nuxt.config.js
nacelle: {
  // ...other Nacelle config,
  shopifyDomain: process.env.SHOPIFY_DOMAIN
}
```

### Add the components to your Nacelle Storefront

There are three components you can add to your Nacelle site: `<loox-product-rating />`, `<loox-product-reviews />`, and `<loox-all-reviews />`.

**Loox Product Rating** will display a star rating for a product along with the number of reviews. Add this component to either `components/ProductDetails.vue` or `components/ProductCard.vue`. It accepts the product object as a prop:

```html
<loox-product-rating :product="product" />
```

**Loox Product Reviews** will display a grid of product reviews for a particular product. Add this component to `pages/products/_handle.vue` or anywhere you use products and pass the product object as a prop:

```html
<loox-product-reviews :product="product" />
```

**Loox All Reviews** will display a grid of all product reviews across multiple products. It accepts no props.

```html
<loox-all-reviews />
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 0.0.3 (2020-10-30)


test/mocks/product.js
export default {
  productId: '4915681296519',
  id: 'pepper-wood-apparel.myshopify.com::bianca-boots::en-us',
  handle: 'bianca-boots',
  locale: 'en-us',
  globalHandle: 'bianca-boots::en-us',
  pimSyncSourceDomain: 'pepper-wood-apparel.myshopify.com',
  title: 'Bianca Boots',
  description:
    '<p>They went out, locking the door of the theatre behind them; and Utterson, once more leaving the servants gathered about the fire in the hall, trudged back to his office to read the two narratives in which this mystery was now to be explained.</p>',
  priceRange: { min: '215.0', max: '215.0', currencyCode: 'USD' },
  availableForSale: true,
  tags: ['filter_color_yellow', 'footwear', 'women'],
  media: [
    {
      id: 'Z2lkOi8vc2hvcGlmeS9JbWFnZVNvdXJjZS84NDYxODQ5NTU5MTc1',
      thumbnailSrc:
        'https://cdn.shopify.com/s/files/1/0344/4362/4583/products/pexels-photo-977908.jpg?v=1587622799&width=100',
      src:
        'https://cdn.shopify.com/s/files/1/0344/4362/4583/products/pexels-photo-977908.jpg?v=1587622799',
      type: 'image',
      altText: 'Bianca Boots'
    }
  ],
  metafields: [],
  variants: [
    {
      id: 'Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0VmFyaWFudC8zMzg5NDA3NDQ4MjgyMw==',
      title: '5',
      price: '215.0',
      priceCurrency: 'USD',
      compareAtPrice: null,
      compareAtPriceCurrency: null,
      swatchSrc: null,
      selectedOptions: [{ name: 'Size', value: '5' }],
      featuredMedia: {
        id: 'Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0SW1hZ2UvMTYyODQ4ODI2OTgzNzU=',
        thumbnailSrc:
          'https://cdn.shopify.com/s/files/1/0344/4362/4583/products/pexels-photo-977908.jpg?v=1587622799&width=100',
        src:
          'https://cdn.shopify.com/s/files/1/0344/4362/4583/products/pexels-photo-977908.jpg?v=1587622799',
        type: 'image',
        altText: null
      },
      sku: null,
      availableForSale: true,
      metafields: [],
      weight: null,
      weightUnit: null,
      priceRules: null
    }
  ],
  indexedAt: 1599762704,
  pimSyncSource: 'shopify',
  pimSyncSourceProductId: 'Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0LzQ5MTU2ODEyOTY1MTk=',
  productType: 'Shoes',
  featuredMedia: {
    id: 'Z2lkOi8vc2hvcGlmeS9JbWFnZVNvdXJjZS84NDYxODQ5NTU5MTc1',
    thumbnailSrc:
      'https://cdn.shopify.com/s/files/1/0344/4362/4583/products/pexels-photo-977908.jpg?v=1587622799&width=100',
    src:
      'https://cdn.shopify.com/s/files/1/0344/4362/4583/products/pexels-photo-977908.jpg?v=1587622799',
    type: 'image',
    altText: 'Bianca Boots'
  },
  vendor: 'Prairie Wind Apparel',
  createdAt: 1587622799
}


test/unit/AllReviewsWidget.test.js
import { mount } from '@vue/test-utils'
import AllReviewsWidget from '~/components/AllReviewsWidget.vue'

const wrapper = mount(AllReviewsWidget)
const component = wrapper.find('.loox-all-reviews')

describe('AllReviewsWidget', () => {
  it('should mount', () => {
    expect(component).not.toBeUndefined()
  })
})


test/unit/ProductRatingWidget.test.js
import { mount } from '@vue/test-utils'
import ProductRatingWidget from '~/components/ProductRatingWidget.vue'
import product from '../mocks/product'

const wrapper = mount(ProductRatingWidget, {
  propsData: { product }
})

const component = wrapper.find('.loox-rating')

describe('ProductRatingWidget', () => {
  it('should mount', () => {
    expect(component).not.toBeUndefined()
  })

  it('should decode the base64 product.pimSyncSourceProductId', () => {
    expect(wrapper.vm.productId).toEqual(product.productId)
  })
})


lib/module.js
const path = require('path')
const { readdirSync } = require('fs')

module.exports = function (moduleOptions) {
  const filename = 'nacelle-loox-plugin.js'
  const options = {
    ...this.options.nacelle,
    ...moduleOptions
  }

  const domain = options.shopifyDomain && options.shopifyDomain.split('.').shift()

  // Add Loox script
  this.options.head.script.push({
    src: `https://loox.io/widget/loox.js?shop=${domain}.myshopify.com`,
    defer: true,
    mode: 'client',
    body: true,
    type: 'text/javascript',
    charset: 'utf-8'
  })

  // Copy components
  const componentsDir = path.resolve(__dirname, 'components')

  for (const file of readdirSync(componentsDir)) {
    this.addTemplate({
      src: path.resolve(__dirname, './components', file),
      fileName: path.join('nacelle', 'integrations', file)
    })
  }

  // Add plugin to nuxt
  this.addPlugin({
    src: path.resolve(__dirname, 'plugin.js'),
    fileName: filename,
    options
  })
}

module.exports.meta = require('../package.json')


example/nuxt.config.js
import path from 'path'
import fs from 'fs-extra'

require('dotenv').config()

export default {
  mode: process.env.BUILD_MODE,
  /*
   ** Headers of the page
   */
  head: {
    title: process.env.npm_package_name || '',
    meta: [
      { charset: 'utf-8' },
      { name: 'viewport', content: 'width=device-width, initial-scale=1' },
      {
        hid: 'description',
        name: 'description',
        content: process.env.npm_package_description || ''
      }
    ],
    link: [{ rel: 'icon', type: 'image/x-icon', href: '/favicon.ico' }]
  },
  /*
   ** Customize the progress-bar color
   */
  loading: { color: '#fff' },

  /*
   ** Global CSS
   */
  css: ['@/assets/global.scss', 'vue-glide-js/dist/vue-glide.css'],

  env: {
    nacelleSpaceID: process.env.NACELLE_SPACE_ID,
    nacelleToken: process.env.NACELLE_GRAPHQL_TOKEN,
    buildMode: process.env.BUILD_MODE,
    contentAssetStorage: process.env.CONTENT_ASSET_STORAGE || ''
  },

  plugins: [
    { src: '~/plugins/nuxt-client-init.js', ssr: false },
    '~/plugins/jsonld'
  ],

  modules: [
    '@nuxtjs/pwa',
    '@nuxtjs/dotenv',
    '@nacelle/nacelle-nuxt-module',
    '@nuxtjs/sitemap',
    '@nuxtjs/axios',
    'nuxt-polyfill',
    'vue-currency-filter/nuxt',
    '~/../lib/module.js'
  ],

  router: {
    middleware: 'cart'
  },

  polyfill: {
    features: [
      {
        require: 'intersection-observer',
        detect: () => 'IntersectionObserver' in window
      }
    ]
  },

  sitemap: {
    gzip: true,
    hostname: 'http://localhost:3000', // When deploying, change this to your production URL
    routes: () => {
      const staticDir = path.resolve(__dirname, './static/data')
      const routes = fs.readJsonSync(`${staticDir}/routes.json`)
      const routesOnly = routes.map(route => route.route)

      return routesOnly
    }
  },

  /*
   ** Nacelle Configuration
   * https://docs.getnacelle.com/nuxt/nuxt-config.html
   */
  nacelle: {
    /* Required Parameters */
    spaceID: process.env.NACELLE_SPACE_ID,
    token: process.env.NACELLE_GRAPHQL_TOKEN,

    /* Optional */

    // Loox
    shopifyDomain: process.env.SHOPIFY_DOMAIN,

    // Google Analytics ID
    // gaID: process.env.NACELLE_GA_ID,

    // Facebook Pixel Tracking ID
    // fbID: process.env.NACELLE_FB_ID,

    // Set the default internationalization locale string for Nacelle to use
    // locale: 'en-us',

    // Customize the route base paths used by Nacelle and Nuxt during generate
    // Learn more in our docs: https://docs.getnacelle.com/nuxt/nuxt-config.html#routeconfig
    // routeConfig: null,

    // Function that can be used for modifying the route array for adding or customizing
    // routes during generate.
    // Learn more in our docs: https://docs.getnacelle.com/nuxt/nuxt-config.html#extendroutes
    // extendRoutes: null,

    // Optional array of data type strings to direct Nacelle to include other data types
    // besides products in search data.
    // searchDataTypes: ['article', 'page', 'blog'],

    // Set to true to bypass fetching data from your space and generating
    // static JSON files.
    // Only set to true if you have previously performed this step at least once.
    // skipPrefetch: process.env.SKIP_PREFETCH === 'true',

    // If you wish to set the Nacelle GraphQL endpoint to something other than the
    // default.
    // customEndpoint: process.env.NACELLE_CUSTOM_ENDPOINT,

    // Set the event tracking endpoint to a URL other than the default
    // tem: process.env.NACELLE_TEM,

    // Customize the build process by creating a new NacelleClient class
    // object.
    // buildClient: null,

    // Enables attempting to fetch data from a user's preferred locale and falling back
    // fetching default locale data.
    isMultiLocale: true
  },

  generate: {
    concurrency: 5,
    done({ errors }, nuxt) {
      nuxt.callHook('generate:done', { nuxt, errors })
    }
  },

  vue: {
    config: {
      devtools: true
    }
  },

  build: {
    postcss: {
      preset: {
        features: {
          customProperties: false
        }
      }
    },
    html: {
      minify: {
        collapseBooleanAttributes: true,
        decodeEntities: true,
        minifyCSS: false,
        minifyJS: false,
        processConditionalComments: true,
        removeEmptyAttributes: true,
        removeRedundantAttributes: true,
        trimCustomFragments: true,
        useShortDoctype: true
      }
    },
    transpile: [
      '@nacelle/nacelle-nuxt-module',
      '@nacelle/nacelle-tools',
      'fuse.js',
      'uuidv4'
    ]
  }
}


module-tests/mocks/product.js
export default {
  id: 'pepper-wood-apparel.myshopify.com::bianca-boots::en-us',
  handle: 'bianca-boots',
  locale: 'en-us',
  globalHandle: 'bianca-boots::en-us',
  pimSyncSourceDomain: 'pepper-wood-apparel.myshopify.com',
  title: 'Bianca Boots',
  description:
    '<p>They went out, locking the door of the theatre behind them; and Utterson, once more leaving the servants gathered about the fire in the hall, trudged back to his office to read the two narratives in which this mystery was now to be explained.</p>',
  priceRange: { min: '215.0', max: '215.0', currencyCode: 'USD' },
  availableForSale: true,
  tags: ['filter_color_yellow', 'footwear', 'women'],
  media: [
    {
      id: 'Z2lkOi8vc2hvcGlmeS9JbWFnZVNvdXJjZS84NDYxODQ5NTU5MTc1',
      thumbnailSrc:
        'https://cdn.shopify.com/s/files/1/0344/4362/4583/products/pexels-photo-977908.jpg?v=1587622799&width=100',
      src:
        'https://cdn.shopify.com/s/files/1/0344/4362/4583/products/pexels-photo-977908.jpg?v=1587622799',
      type: 'image',
      altText: 'Bianca Boots'
    }
  ],
  metafields: [],
  variants: [
    {
      id: 'Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0VmFyaWFudC8zMzg5NDA3NDQ4MjgyMw==',
      title: '5',
      price: '215.0',
      priceCurrency: 'USD',
      compareAtPrice: null,
      compareAtPriceCurrency: null,
      swatchSrc: null,
      selectedOptions: [{ name: 'Size', value: '5' }],
      featuredMedia: {
        id: 'Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0SW1hZ2UvMTYyODQ4ODI2OTgzNzU=',
        thumbnailSrc:
          'https://cdn.shopify.com/s/files/1/0344/4362/4583/products/pexels-photo-977908.jpg?v=1587622799&width=100',
        src:
          'https://cdn.shopify.com/s/files/1/0344/4362/4583/products/pexels-photo-977908.jpg?v=1587622799',
        type: 'image',
        altText: null
      },
      sku: null,
      availableForSale: true,
      metafields: [],
      weight: null,
      weightUnit: null,
      priceRules: null
    }
  ],
  indexedAt: 1599762704,
  pimSyncSource: 'shopify',
  pimSyncSourceProductId: 'Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0LzQ5MTU2ODEyOTY1MTk=',
  productType: 'Shoes',
  featuredMedia: {
    id: 'Z2lkOi8vc2hvcGlmeS9JbWFnZVNvdXJjZS84NDYxODQ5NTU5MTc1',
    thumbnailSrc:
      'https://cdn.shopify.com/s/files/1/0344/4362/4583/products/pexels-photo-977908.jpg?v=1587622799&width=100',
    src:
      'https://cdn.shopify.com/s/files/1/0344/4362/4583/products/pexels-photo-977908.jpg?v=1587622799',
    type: 'image',
    altText: 'Bianca Boots'
  },
  vendor: 'Prairie Wind Apparel',
  createdAt: 1587622799
}
